src/RecyclableMemoryStream.cs

```diff
@@ -619,11 +623,126 @@ async Task CopyToAsyncImpl(CancellationToken ct)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), $"{nameof(count)} must be non-negative.");
+            }
+
+            byte[] buffer = this.bufferWriterTempBuffer;
+            if (buffer != null)
+            {
+                if (count > buffer.Length)
+                {
+                    throw new InvalidOperationException($"Cannot advance past the end of the buffer, which has a size of {buffer.Length}.");
+                }
+
+                this.Write(buffer, 0, count);
+                this.ReturnTempBuffer(buffer);
+                this.bufferWriterTempBuffer = null;
+            }
+            else
+            {
+                long bufferSize = this.largeBuffer == null
+                    ? this.memoryManager.BlockSize - this.GetBlockAndRelativeOffset(this.position).Offset
+                    : this.largeBuffer.Length - this.position;
+
+                if (count > bufferSize)
+                {
+                    throw new InvalidOperationException($"Cannot advance past the end of the buffer, which has a size of {bufferSize}.");
+                }
+
+                this.position += count;
+                this.length = Math.Max(this.position, this.length);
+            }
+        }
+
+        private void ReturnTempBuffer(byte[] buffer)
+        {
+            if (buffer.Length == this.memoryManager.BlockSize)
+            {
+                this.memoryManager.ReturnBlock(buffer, this.id, this.tag);
+            }
+            else
+            {
+                this.memoryManager.ReturnLargeBuffer(buffer, this.id, this.tag);
+            }
+        }
+
+        /// <inheritdoc/>
+        /// <remarks>
+        /// IMPORTANT: Calling Write(), GetBuffer(), TryGetBuffer(), Seek(), GetLength(), Advance(),
+        /// or setting Position after calling GetMemory() invalidates the memory.
+        /// </remarks>
+        public Memory<byte> GetMemory(int sizeHint = 0) => this.GetWritableBuffer(sizeHint);
+
+        /// <inheritdoc/>
+        /// <remarks>
+        /// IMPORTANT: Calling Write(), GetBuffer(), TryGetBuffer(), Seek(), GetLength(), Advance(),
+        /// or setting Position after calling GetMemory() invalidates the memory.
+        /// </remarks>
+        public Span<byte> GetSpan(int sizeHint = 0) => this.GetWritableBuffer(sizeHint);
+
+        /// <summary>
+        /// When callers to GetSpan() or GetMemory() request a buffer that is larger than the remaining size of the current block
+        /// this method return a temp buffer. When Advance() is called, that temp buffer is then copied into the stream.
+        /// </summary>
+        private ArraySegment<byte> GetWritableBuffer(int sizeHint)
+        {
+            this.CheckDisposed();
+            if (sizeHint < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(sizeHint), $"{nameof(sizeHint)} must be non-negative");
+            }
+
+            sizeHint = Math.Max(sizeHint, 1);
```

Comment:
Given that, at this point, `sizeHint` can't be less than `0`, this line of code is functionally equivalent to:

```CSharp
if (sizeHint == 0)
{
    sizeHint = 1;
}
```

With a better performance footprint.